### PR TITLE
Prime cache for deb "installer"

### PR DESCRIPTION
**Pull request by wli3**
image

#6370

### CODE REVIEW

**Comment by wli3** [comment]
@bleroy

**Comment by wli3** [comment]
@dotnet-bot Test RHEL7.2 x64 Release Build

File: packaging/deb/postinst [hunk 1]
```diff
@@ -1,3 +1,13 @@
 #!/usr/bin/env sh
 echo "This software may collect information about you and your use of the software, and send that to Microsoft."
 echo "Please visit http://aka.ms/dotnet-cli-eula for more information."
+
+# Run 'dotnet new' to trigger the first time experience to initialize the cache
+echo ."NET Core Tools Telemetry
+
+The .NET Core Tools include a telemetry feature (https://github.com/dotnet/cli/pull/2145) that collects usage information. It’s important that the .NET Team understands how the tools are being used so that we can improve them.
```
Comment (eerhardt) [inline packaging/deb/postinst]: Do we link to the PR https://github.com/dotnet/cli/pull/2145 in all the other notices?  That seems strange because we can and probably will add more telemetry related PRs in the future.  So just calling this one out seems misleading.

File: build/package/Installer.DEB.targets [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Target Name="SetupDebProps"
-          DependsOnTargets="Init">
+      DependsOnTargets="Init">
```
Comment (eerhardt) [inline build/package/Installer.DEB.targets]: These kinds of changes shouldn't happen.  Lining the attributes up vertically is very common, and helps in reading these files.

File: packaging/deb/postinst (hunk 1 above)
Comment (nguerrera) [inline packaging/deb/postinst]: I don't think we should ever link to a PR from the product. It should be an aka.ms or fwlink and it should land in docs not github. 

The notices should be the same across platforms too.

Comment (wli3) [inline packaging/deb/postinst]: This is suggested by @bleroy, the link to the PR is on our docs already
https://docs.microsoft.com/en-us/dotnet/articles/core/tools/telemetry

The MD file is there
https://github.com/dotnet/docs/edit/master/docs/core/tools/telemetry.md

File: build/package/Installer.DEB.targets (hunk 2 above)
Comment (wli3) [inline build/package/Installer.DEB.targets]: No strong opinion, this file has a lot of extra whitespace so I run xml format over the file to fix them.

I will revert it back

Comment (wli3) [inline build/package/Installer.DEB.targets]: And ideally if https://github.com/dotnet/codeformatter can work with our repo, this kind of conversion help improve the formatting config file instead of a one time fix. And I am working on it slowly

File: packaging/deb/postinst (hunk 1 above)
Comment (wli3) [inline packaging/deb/postinst]: So I will keep only keep the aka link at the end

Comment (wli3) [inline packaging/deb/postinst]: This is fixed, please see the latest file

File: build/package/Installer.DEB.targets (hunk 2 above)
Comment (wli3) [inline build/package/Installer.DEB.targets]: This is fixed

**Comment by wli3** [comment]
@dotnet-bot Test RHEL7.2 x64 Release Build
@dotnet-bot Test Ubuntu x64 Release Build

**Comment by livarcocc** [comment]
@ManishJayaswal @Pilchie for approval.

File: packaging/deb/postinst (hunk 1 above)
Comment (wli3) [inline packaging/deb/postinst]: #6448 use the original one back, and will send a separate pull request to get text sign off specifically

File: build/package/Installer.DEB.proj [hunk 3]
```diff
@@ -10,7 +10,7 @@
                             TestDebuild;
                             BuildSdkDeb;
                             TestSdkDeb;"
-          Condition=" '$(OSName)' == 'ubuntu' " 
+          Condition=" '$(OSName)' == 'ubuntu' "
```
Comment (nguerrera) [inline build/package/Installer.DEB.proj]: Put whitespace only diffs in separate commits in the future.

**Comment by Pilchie** [comment]
What is the customer scenario here?  Should this have an ask mode template?

**Comment by wli3** [comment]
**Customer scenario**

In the last step of customer installing CLI debain package, `sudo dpkg -i dotnet-sdk-VERSION` , the installer should invoke a CLI command to trigger the first run experience. It will extract the NuGet cache to local machine. This is important for Visual Studio. The Windows and macOS counterpart has been checked in, this is for Linux Ubuntu/Debian distro.

The same time the telemetry message should be shown in the installer, because the user won't get it in the first run.

**Bugs this fixes**

[CLI #6370](https://github.com/dotnet/cli/issues/6370)

**Workarounds, if any**

N/A

**Risk**

The user needs to wait for NuGet cache to be extracted at the first time CLI is invoked.

There is no feature parity between Windows and Ubuntu

**Performance impact**

N/A

**Root cause analysis**

N/A

**How was the bug found?**

N/A

**Comment by wli3** [comment]
@dotnet-bot test Linux x64 Release Build

**Comment by wli3** [comment]
@dotnet-bot test Ubuntu16.04 x64 Debug Build
@dotnet-bot test Ubuntu16.10 x64 Debug Build

**Comment by livarcocc** [comment]
@dotnet-bot Test Ubuntu x64 Release Build

**Comment by wli3** [comment]
The first commit does have the same people with https://github.com/dotnet/cli/pull/6500

Add the second commit fixed it 157ee13778553edada4c1a933a3cdee23e5b4303
@livarcocc

**Comment by wli3** [comment]
@dotnet-bot test Debian8.2 x64 Debug Build
@dotnet-bot test RHEL7.2 x64 Release Build 
@dotnet-bot test Ubuntu16.04 x64 Debug Build

**Comment by wli3** [comment]
@dotnet-bot  test Windows_NT x86 Debug Build

**Comment by omajid** [comment]
Sorry for jumping in here very late. I suppose this same patch (or a derivative) will apply to RPMs too. I have a few questions:

> This is important for Visual Studio. 

Can you elaborate? Visual Studio doesn't work on Linux so this can't possibly impact Visual Studio, can it?

> The same time the telemetry message should be shown in the installer

If this is a non-interactive installation, messages from the package install may not be shown to the user. And since the first run cache exists on subsequent runs, the message is never shown again.

The `dotnet new` invocation is a surprising implementation step that works differently under different scenarios:

- If there are 2 (normal) users on machine, only one gets the cache when .NET Core is installed.
- If root installed packages (no `sudo` used in installation, possibly via some GUI software center) root gets nuget cache, no one else does.

Packages touching a user's files (in this case under `/home/$SUDO_USER`)  is generally a no-no: https://stackoverflow.com/questions/3140886/debian-packaging-deploying-files-to-the-user-home-directory
